src/main/kotlin/bsu/cc/parser/XlsxParser.kt

```diff
@@ -0,0 +1,152 @@
+import java.lang.IllegalArgumentException
+import java.lang.IllegalStateException
+import java.lang.reflect.Type
+import java.util.*
+
+
+
+fun readWorkbook(fileName: String): XSSFWorkbook {
+    val file = FileInputStream(File(fileName))
+    return XSSFWorkbook(file)
+}
+
+fun <T> sheetToDataClasses(
+        sheet: Sheet,
+        dataProducer: (namedRowMap: Map<String, Cell>) -> T,
+        ignoreDuplicateHeaders: Boolean = false
+): Sequence<T> {
+    val headerMap = sheetHeaderMap(sheet)
+    fun translatedProducer(indexedRowMap: Map<Int, Cell>): T {
+        return dataProducer(indexedToNamedRowMap(indexedRowMap, headerMap, ignoreDuplicateHeaders))
+    }
+    return sheetToDataClasses(sheet, ::translatedProducer, true)
+}
+
+fun <T> sheetToDataClasses(
+        sheet: Sheet,
+        dataProducer: (indexedRowMap: Map<Int, Cell>) -> T,
+        excludeHeader: Boolean? = null
+): Sequence<T> {
+    return sheet.iterator().asSequence().withIndex().filter { (_, row) ->
+        row.cellIterator().asSequence().filter{cell -> cell.cellType != CellType.BLANK}.toList().isNotEmpty() //Ignore blank rows
+    }.filter { (index, _) ->
+        excludeHeader == null || !(excludeHeader && index == 0)
+    }.map { (_, row) ->
+        rowToDataClass(row, dataProducer)
+    }
+}
+
+private fun indexedToNamedRowMap(
+        indexedRowMap: Map<Int, Cell>,
+        colNameMap: Map<Int, String>,
+        ignoreDuplicateColumns: Boolean = false
+): Map<String, Cell> {
+    val output = HashMap<String, Cell>()
+    indexedRowMap.keys.forEach { index ->
+        val colName = colNameMap[index]
+        val rowCell = indexedRowMap[index]
+        if(colName == null) {
+            throw IllegalArgumentException("Accessing column index out of bounds of headers")
+        }
+        if(rowCell == null) {
+            throw IllegalStateException("Key not in map") //This should never happen, but needed for smartcast
+        }
+        if(!output.containsKey(colName)) {
+            output[colName] = rowCell
+        } else if(!ignoreDuplicateColumns) {
+            throw IllegalArgumentException("Column map contains duplicate column names")
+        }
+    }
+    return output
+}
+
+fun sheetHeaderMap(sheet: Sheet): Map<Int, String> {
+    val output = HashMap<Int, String>()
+    val headerRow = sheet.first()
+    //Can't use foreach iterator because it skips empty columns, which causes column indices to be wrong
+    (0 until headerRow.lastCellNum).forEach{index ->
+        if(headerRow.getCell(index) != null) {
+            val cell = headerRow.getCell(index)
+            if(cell != null) {
+                if(cell.cellType != CellType.STRING) {
+                    throw IllegalArgumentException("Sheet has non-string in header")
+                }
+                output[index] = cell.stringCellValue
+            }
+        }
+    }
+    return output
+}
+
+fun <T> rowToDataClass(
+        row: Row,
+        dataProducer: (rowMap: Map<Int, Cell>) -> T
+): T {
```

Comment:
This would be another case where `= dataProducer(rowToCellMap(row))` would be more idiomatic.